rfcs/0191-worker-manager-launch-configs.md

```diff
@@ -0,0 +1,122 @@
+## Details
+
+New entity will be introduced to worker-manager - **Launch Configuration**.
+Each worker pool will have zero or more **launch configurations** associated with it.
+
+Each **launch configuration** will be assigned a unique ID (hash of its properties), stored in the database and will be immutable (except for the status flag).
+
+Generated ID will include the following properties:
+
+* top-level config properties (`minCapacity`, `maxCapacity`, `scalingRatio`, `lifecyle`)
+* launch configuration itself
+
+During updates of the worker pool configurations, all existing launch configurations that are not present in the new configuration will be marked as `archived`. Launch configurations that have same unique ID (hash) will be kept active. All other will be created as new launch configurations and marked as active.
+
+### Launch Configuration weight
+
+Provisioned workers would be associated with a specific launch configuration (`worker.launchConfigurationId`).
+This will allow us to to know how many workers with this launch configuration were successfully registered and claimed work.
+
+Each launch configuration will have a dynamic `weight` property that will be adjusted automatically based on the following events and metrics:
+
+* total number of successful worker provisioning attempts / registrations
+* total number of failed worker provisioning attempts / registrations
+* has any worker claimed a task
+* fail-to-success ratio over a specific time period (i.e. last hour)
+* number of non-stopped workers currently running
+
+Weight will be used to determine the likelihood of selecting a specific launch configuration for provisioning.
+
+This will be calculated at the provisioning time.
+
+### Making worker-manager extensible
+
+Worker-manager will publish additional events to Pulse to allow external systems to react:
+
+* `launch-configuration-created`
+* `launch-configuration-archived`
+* `launch-configuration-paused`
+* `launch-configuration-resumed`
+* `worker-registered`
+* `worker-failure` (provisioning or starting up failure, will include `workerPoolId` and `launchConfigurationId`)
+
+New API endpoints will be introduced:
+
+* `workerManager.getLaunchConfigs(workerPoolId)` - to retrieve all launch configurations with their statuses
+* `workerManager.getLaunchConfig(workerPoolId, launchConfigId)` - to retrieve specific launch configuration
+* `workerManager.pauseLaunchConfig(workerPoolId, launchConfigId)` - to deactivate specific active launch configuration (time bound)
+* `workerManager.resumeLaunchConfig(workerPoolId, launchConfigId)` - to resume specific active launch configuration
+
+Last two endpoints would allow external systems to pause/resume specific launch configurations based on their own criteria.
+This might be useful when dynamic adjustment of weights is not enough to prevent provisioning workers in a specific region.
+
+Existing endpoints will continue to accept the same payload as before for backward compatibility.
```

Comment:
I suspect that the create/updateWorkerPool methods will need to return a set of launch configs, too?

Reply:
Yes, right. Since it would be a breaking change anyway. 

Worth keeping in mind that tc-admin/ciadmin might need to be patched.